spirv_glsl.cpp

```diff
@@ -3348,100 +3305,285 @@ void CompilerGLSL::emit_extension_workarounds()
 {
 	const auto TYPES = { "int",   "ivec2", "ivec3", "ivec4", "uint",   "uvec2", "uvec3", "uvec4",
 		                 "float", "vec2",  "vec3",  "vec4",  "double", "dvec2", "dvec3", "dvec4" };
-	// GL_KHR_shader_subgroup_ballot
-	if (!options.vulkan_semantics && declared_KHR_shader_subgroup_ballot)
+
+	if (!options.vulkan_semantics)
 	{
-		statement("#ifdef GL_NV_shader_thread_group");
-		statement("uvec4 subgroupBallot(bool v) { return uvec4(ballotThreadNV(v), 0u, 0u, 0u); }");
-		statement("#elif defined(GL_ARB_shader_ballot)");
-		statement("uvec4 subgroupBallot(bool v) { return uvec4(unpackUint2x32(ballotARB(v)), 0u, 0u); }");
-		statement("#endif");
+		using Supp = KHR_shader_subgroupSupportHelper;
 
+		Supp::Result result = shader_subgroup_supporter.resolve();
 
-		statement("#ifdef GL_NV_shader_thread_shuffle");
+		if (shader_subgroup_supporter.isFeatureRequested(Supp::SubgroupMask))
+		{
+			auto exts = shader_subgroup_supporter.getCandidatesForFeature(Supp::SubgroupMask, result);
 
-		for (const char *t : TYPES)
-			statement(string(t) + " subgroupBroadcastFirst(" + t +
-			          " value) { return shuffleNV(value, 0, gl_WarpSizeNV); }");
-		for (const char *t : TYPES)
-			statement(string(t) + " subgroupBroadcast(" + t +
-			          " value, uint id) { return shuffleNV(value, id, gl_WarpSizeNV); }");
+			for (uint32_t i = 0u; i < exts.size(); ++i)
+			{
+				const Supp::Candidate e = exts[i];
+				const std::string name = Supp::getExtensionName(e);
 
-		statement("#elif defined(GL_ARB_shader_ballot)");
+				statement(i == 0u ? "#if" : "#elif", " defined(", name, ")");
+				switch (e)
+				{
+				case Supp::NV_shader_thread_group:
+					statement("#define gl_SubgroupEqMask uvec4(gl_ThreadEqMaskNV, 0u, 0u, 0u)");
```

Comment:
I see a problem here, one that can be easily solved.

If `KHR_shader_subgroup_X` is defined (available) then you shouldn't emit any workarounds for `KHR_shader_subgroup_X`

This you could solve by actually having the `KHR` extensions in the candidate set, and have them take precedence even over the NV extensions.

Reply:
fixed